lib/jobs/ssh-job.js

```diff
@@ -0,0 +1,207 @@
+                    result.cmd = commandData.command; //for parser matching
+                    return results.concat([result]);
+                });
+            }, []);
+        })
+        .then(self.handleResponse.bind(self))
+        .then(function() {
+            self._done();
+        })
+        .catch(self._done.bind(self));
+    };
+
+    SshJob.prototype.handleResponse = function(results) {
+        var self = this;
+
+        logger.debug("Received remote command output from node.", {
+            id: self.nodeId,
+        });
+
+        var catalogTasks = _.filter(results, function(result) {
+            logger.debug('task result', {data:result});
+            return  _.has(result, 'catalogOptions') && result.catalogOptions;
+        });
+
+        return self.catalogUserTasks(catalogTasks)
+        .catch(function(err) {
+            logger.error("Job error processing catalog output.", {
+                error: err,
+                id: self.nodeId,
+                taskContext: self.context
+            });
+            throw err;
+        });
+    };
+
+    SshJob.prototype.catalogUserTasks = function(tasks) {
+        var self = this;
+
+        return parser.parseUnknownTasks(tasks)
+        .spread(function() {
+            return Promise.map(Array.prototype.slice.call(arguments), function(result) {
+                if (result.error) {
+                    logger.error("Failed to parse data for " +
+                        result.source + ', ' + result.error,
+                        { error: result });
+                } else if (result.store) {
+                    return waterline.catalogs.create({
+                        node: self.nodeId,
+                        source: result.source || 'unknown',
+                        data: result.data
+                    });
+                } else {
+                    logger.info("Catalog result for " + result.source +
+                        " has not been marked as significant. Not storing.");
+                }
+            });
+        });
+    };
+
+    SshJob.prototype.sshExec = function(command, sshSettings, sshClient) {
+        var self = this;
+        return new Promise(function(resolve, reject) {
+
+            var ssh = sshClient;
+            var sshData = {};
+            ssh.on('ready', function() {
+                ssh.exec(command, function(err, stream) {
+                    if (err) { reject(err); }
+                    stream.on('close', function(code) {
+                        sshData.exitCode = code;
+                        ssh.end();
+                    }).on('data', function(data) {
+                        sshData.stdout = ( sshData.stdout || '' ) + data.toString();
+                    }).stderr.on('data', function(data) {
+                        sshData.stderr = ( sshData.stderr || '' ) + data.toString();
+                    });
+                });
+            }).on('close', function() {
```

Comment:
From the user guide from [ssh2](https://www.npmjs.com/package/ssh2), the `close` event will pass a argument `hadError` to the callback, so here needs to reject the promise if `hadError` is positive.

Meanwhile, there is another event 'error`, this event need be better to be caught to reject promise and print the error info to assist debugging.